cedar-lean/Cedar/Validation/RequestEntityValidator.lean

```diff
@@ -0,0 +1,157 @@
+  if instanceOfEntityType request.principal reqty.principal then
+    if request.action == reqty.action then
+      if instanceOfEntityType request.resource reqty.resource then
+        if instanceOfType request.context (.record reqty.context) then true
+        else false
+      else false
+    else false
+  else false
+/--
+For every entity in the store,
+1. The entity's type is defined in the type store.
+2. The entity's attributes match the attribute types indicated in the type store.
+3. The entity's ancestors' types are consistent with the ancestor information
+   in the type store.
+-/
+
+def instanceOfEntitySchema (entities : Entities) (ets : EntitySchema) : EntityValidationResult :=
+entities.toList.forM (λ (uid, data) => match ets.find? uid.ty with
+  | .some entry => if instanceOfType data.attrs (.record entry.attrs) then
+                    if data.ancestors.all (λ ancestor => entry.ancestors.contains ancestor.ty) then .ok ()
+                   else .error (.typeError s!"entity ancestors inconsistent with type store information")
+                  else .error (.typeError "entity attributes do not match type store")
+  | _ => .error (.typeError "entity type not defined in type store"))
+
+/--
+For every action in the entity store, the action's ancestors are consistent
+with the ancestor information in the action store.
+-/
+def instanceOfActionSchema (entities : Entities) (as : ActionSchema) : EntityValidationResult :=
+as.toList.forM (λ (uid, data) => match entities.find? uid with
+  | .some entry => if data.ancestors == entry.ancestors
+                    then .ok ()
+                    else .error (.typeError "action ancestors inconsistent with type store information")
+  | _ => .error (.typeError s!"action type {uid.eid} not defined in type store"))
+
+def requestMatchesEnvironment (env : Environment) (request : Request) : Bool := instanceOfRequestType request env.reqty
+
+def validateRequest (schema : Schema) (request : Request) : RequestValidationResult :=
+  if ((schema.toEnvironments.any (requestMatchesEnvironment · request)))
+  then .ok ()
+  else .error (.typeError "request could not be validated in any environment")
+
+def entitiesMatchEnvironment (env : Environment) (entities : Entities) : EntityValidationResult :=
+instanceOfEntitySchema entities env.ets >>= λ _ => instanceOfActionSchema entities env.acts
+
+def actionSchemaEntryToEntityData (ase : ActionSchemaEntry) : EntityData := {
+  ancestors := ase.ancestors,
+  attrs := Map.empty
+}
+
+/--
+Update the entity schema with the entities created for action schema entries.
+This involves the construction of the ancestor information for the associated types
+by inspecting the concrete hierarchy.
+-/
+def updateSchema (schema : Schema) (actionSchemaEntities : Entities) : Schema :=
+  let uniqueTys := Set.make (actionSchemaEntities.keys.map (·.ty)).elts
+  let newEntitySchemaEntries := uniqueTys.elts.map (λ ty =>
```

Comment:
The inlined lambda here is complex.  Consider moving it to a separate `where` clause.